### PR TITLE
Fix showing custom masks

### PR DESCRIPTION
**Pull request by cquiroz**
This is a fix to show the custom mask properly on the steps table for GMOS:

image

### CODE REVIEW

File: modules/seqexec/server/src/main/scala/seqexec/server/gmos/GmosNorth.scala [hunk 1]
```diff
@@ -30,18 +30,26 @@ final case class GmosNorth[F[_]: Concurrent: Timer: Logger] private (
...
+  ) extends Gmos[F, NorthTypes](c,
   new SiteSpecifics[NorthTypes] {
     def extractFilter(config: CleanConfig): Either[ConfigUtilOps.ExtractFailure, NorthTypes#Filter] =
       config.extractInstAs[NorthTypes#Filter](FILTER_PROP)
     def extractDisperser(config: CleanConfig): Either[ConfigUtilOps.ExtractFailure, GmosNorthType.DisperserNorth] =
       config.extractInstAs[NorthTypes#Disperser](DISPERSER_PROP)
     def extractFPU(config: CleanConfig): Either[ConfigUtilOps.ExtractFailure, GmosNorthType.FPUnitNorth] =
       config.extractInstAs[NorthTypes#FPU](FPU_PROP_NAME)
+    def extractCustomFPU(config: CleanConfig): Either[ConfigUtilOps.ExtractFailure, String] =
+      config.extractInstAs[String]("fpuCustomMask")
```
Comment (rpiaggio) [inline modules/seqexec/server/src/main/scala/seqexec/server/gmos/GmosNorth.scala:42]: Should we put the property name in a constant for consistency with the other ones?

File: modules/seqexec/server/src/main/scala/seqexec/server/gmos/GmosSouth.scala [hunk 2]
```diff
@@ -38,11 +38,18 @@ final case class GmosSouth[F[_]: Concurrent: Timer: Logger] private (
       config.extractInstAs[SouthTypes#Filter](FILTER_PROP)
     def extractDisperser(config: CleanConfig): Either[ConfigUtilOps.ExtractFailure, GmosSouthType.DisperserSouth] =
       config.extractInstAs[SouthTypes#Disperser](DISPERSER_PROP)
+    def extractCustomFPU(config: CleanConfig): Either[ConfigUtilOps.ExtractFailure, String] =
+      config.extractInstAs[String]("fpuCustomMask")
```
Comment (rpiaggio) [inline modules/seqexec/server/src/main/scala/seqexec/server/gmos/GmosSouth.scala:42]: Same here, we could reference the same constant.

File: modules/seqexec/server/src/main/scala/seqexec/server/gmos/GmosSouth.scala [hunk 3]
```diff
@@ -38,11 +38,18 @@ final case class GmosSouth[F[_]: Concurrent: Timer: Logger] private (
...
     def extractFPU(config: CleanConfig): Either[ConfigUtilOps.ExtractFailure, GmosSouthType.FPUnitSouth] =
       config.extractInstAs[SouthTypes#FPU](FPU_PROP_NAME)
     def extractStageMode(config: CleanConfig): Either[ConfigUtilOps.ExtractFailure, GmosSouthType.StageModeSouth] =
       config.extractInstAs[SouthTypes#GmosStageMode](STAGE_MODE_PROP)
-    def isCustomFPU(config: CleanConfig): Boolean = extractFPU(config).map(_.isCustom()).getOrElse(false)
+    def isCustomFPU(config: CleanConfig): Boolean =
```
Comment (rpiaggio) [inline modules/seqexec/server/src/main/scala/seqexec/server/gmos/GmosSouth.scala:47]: Seems to be the same logic at both sites, can it be moved to `SiteSpecifics`?

File: modules/seqexec/web/client/src/main/scala/seqexec/web/client/model/StepItems.scala [hunk 4]
```diff
@@ -38,45 +37,40 @@ object StepItems {
...
-        fpuL = if (isNoneFPU(i) || mode === FPUMode.BuiltIn) instrumentFPUO
-        else instrumentFPUCustomMaskO
-        fpu <- fpuL.getOption(s)
-      } yield fpuNameMapper(i)(fpu).getOrElse(fpu)
+      (i, instrumentFPUO.getOption(s), instrumentFPUCustomMaskO.getOption(s)) match {
+        case (Instrument.GmosS | Instrument.GmosN | Instrument.F2, Some("CUSTOM_MASK"), c) => c
+        case (Instrument.GmosS | Instrument.GmosN | Instrument.F2, None, c @ Some(_))      => c
+        case (_, Some(b), _)                                                               => fpuNameMapper(i)(b)
+        case _                                                                             => none
+      }
```
Comment (rpiaggio) [inline modules/seqexec/web/client/src/main/scala/seqexec/web/client/model/StepItems.scala:73]: If I understand correctly, we are not defaulting to `BuiltIn` anymore, is that the fix?

**Review by rpiaggio — APPROVED** [review]
LGTM

File: modules/seqexec/web/client/src/main/scala/seqexec/web/client/model/StepItems.scala (hunk 4 above)
Comment (cquiroz) [reply modules/seqexec/web/client/src/main/scala/seqexec/web/client/model/StepItems.scala:73]: There are several options how the OT presents the sequence depending on whether there are iterators and what not.
This basically tries to find if there is a custom mask present first and if not it goes the build in
There is a flag to indicate whether we use custom or built in and it was used previously
but it turns out it is unreliable and there are cases where the flag indicates we have a built in mask
while in reality it is custom

File: modules/seqexec/server/src/main/scala/seqexec/server/gmos/GmosSouth.scala (hunk 3 above)
Comment (cquiroz) [reply modules/seqexec/server/src/main/scala/seqexec/server/gmos/GmosSouth.scala:47]: The logic looks the same but they are different as they act on different types representing FPU North/South
`extractCustomFPU` can be made common though.

File: modules/seqexec/server/src/main/scala/seqexec/server/gmos/GmosNorth.scala [hunk 5]
```diff
@@ -41,7 +41,12 @@ final case class GmosNorth[F[_]: Concurrent: Timer: Logger] private (
     def extractStageMode(config: CleanConfig): Either[ConfigUtilOps.ExtractFailure, GmosNorthType.StageModeNorth] =
       config.extractInstAs[NorthTypes#GmosStageMode](STAGE_MODE_PROP)
     val fpuDefault: GmosNorthType.FPUnitNorth = FPU_NONE
-    def isCustomFPU(config: CleanConfig): Boolean = extractFPU(config).map(_.isCustom()).getOrElse(false)
+    def isCustomFPU(config: CleanConfig): Boolean =
+      (extractFPU(config), extractCustomFPU(config)) match {
+        case (Right(builtIn), _) if builtIn.isCustom() => true
+        case (_, Right(_))                             => true
```
Comment (swalker2m) [inline modules/seqexec/server/src/main/scala/seqexec/server/gmos/GmosNorth.scala:47]: I would have expected the previous version to work too.  How are the sequence steps generated?  There is an option that gives you the full configuration at each step and the FPU should be correct there.  If you want to send me a sequence where it is wrong, I'll take a look.

Comment (swalker2m) [reply modules/seqexec/server/src/main/scala/seqexec/server/gmos/GmosNorth.scala:47]: Thanks, I looked at the sequence you sent.  It's clearly a bug in the `ocs` code but this looks like it should work around it.  :-/

**Review by jluhrs — APPROVED** [review]
LGTM

**Pull request merged**